csrc/multidevice/utils.cpp

```diff
@@ -13,22 +13,93 @@
 #include <multidevice/lower_communication.h>
 #include <multidevice/utils.h>
 #include <ops/all_ops.h>
+#include <root_domain_map.h>
 #include <scheduler/utils.h>
 
 #include <c10/util/irange.h>
 
 namespace nvfuser {
 namespace {
 
-std::vector<IterDomain*> getShardedIterDomains(TensorView* tv) {
-  std::vector<IterDomain*> sharded_ids;
+const std::unordered_set<IterDomain*> getShardedIterDomains(TensorView* tv) {
+  std::unordered_set<IterDomain*> sharded_ids;
   std::copy_if(
       tv->getLeafDomain().begin(),
       tv->getLeafDomain().end(),
-      std::back_inserter(sharded_ids),
+      std::inserter(sharded_ids, sharded_ids.begin()),
       [](auto id) { return id->isDeviceDim(); });
   return sharded_ids;
 }
+
+// Returns whether a IterDomain in a TensorView can be read/written
+// contiguously. when a IterDomain's sharding is flipped. This occurs when the
+// IterDomain being unsharded/sharded is the outermost in the allocation domain.
+// e.g. [DIDx(i0), r1, i2, i3]
+// i0 => true, outermost axis
+// i1 => true, reduction axis are not allocated
+// i2 => true, reduction and device axis are not allocated
+// i3 => false, i2 is allocated.
+bool isContiguousShard(TensorView* tv, IterDomain* changed_id) {
+  bool not_allocated = true;
+  for (auto id : tv->getLeafDomain()) {
+    if (id == changed_id) {
+      return not_allocated;
+    }
+    not_allocated = not_allocated && (id->isDeviceDim() || id->isReduction());
+  }
+  NVF_ERROR(
+      false,
+      "Id",
+      changed_id->toString(),
+      " is not in TensorView ",
+      tv->toString());
+  return false;
+}
+
+// For a sharding expression, either a set or reduce, returns root IDs
```

Comment:
Question about naming: I saw sometimes we say resharding expressions (or ops) vs sharding expressions. Do they have different meanings or are used interchangeably?

Reply:
I used them interchangeably. I can make this more consistent if there is a preferred naming for nvfuser?

Reply:
I don't have a strong preference. Resharding seems to be more precise -- it says the fact that the sharding is changed after this op. Anyhow, I worried about consistency more than which name.

Reply:
Yes, I updated sharding to resharding and went through the PR to make certain the naming was consistent.